### ISSUE TITLE
Follow Semantic Versioning

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#2387]
### Describe the project you are working on
This applies to all projects.

### Describe the problem or limitation you are having in your project
Because new features are also introduced in Patch releases, there never is a "most stable Godot version".

Right now it is impossible to know at a glance if a new Godot patch release is going to bring up conflicts or not when used with existing projects. The (usually hobbyist) user has to read carefully through all the release notes, check issues and thoroughly test their project to know a new release is save to use. Even though patch releases containing 95% desperately needed bug fixes. 
As it is evident through regular surveys, Godot's userbase by enlarge consists of hobbyists, who are by definition only part time users. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
If instead Godot would strictly follow [Semantic Versioning](https://semver.org/) and minor releases would happen more frequently, any user would immediately know at a glance what they are getting.
This would especially help users who do not follow the Godot development on a daily basis (the vast majority of Godot users). They could profit from the latest fixes, being able to use the new patch knowing "things have been fixed", knowing they are on the most stable version. Their user experience has only improved, thanks to the patch, other than that, nothing fundamentally has changed. They would not have to spend their limited time plowing through patch release notes, spend sparse development time testing how new features (they might not even need) interfere with their project, but rather could just get to work and benefit from the most stable Godot experience there is.

The distinction between minor and patch releases becomes more meaningful.
"Do I want this new feature or is stability currently more important?" would be a much more meaningful choice, easily made by just looking at the version number of a new release. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
![image](https://user-images.githubusercontent.com/47016402/109652420-112a6500-7b60-11eb-8c6a-bcf83c02ef73.png)

- When the major version number changes, it means there have been changes so substantial, they break forward and backward compatibility. Users need either to manually convert their projects or use an auto-convertion tool if they want to use existing projects with the new version.
- When the minor version number changes, it means this release comes with new features while keeping forward compatibility. Users can weigh the relevance of the new features for their projects against the reduced stability compared to the previous version. Some feature changes may also break backward compatibility, which should be clearly marked in the release notes.
- When the patch version number changes, it means only bug fixes have been added. The latest patch version is therefore always the most stable Godot version.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
It's about versioning, so no.

### Is there a reason why this should be core and not an add-on in the asset library?
It's about versioning of the core engine.

**Comment by Calinou** [comment]
See also https://github.com/godotengine/godot-proposals/issues/2095.

Many of the core contributors are still not 100% convinced that following semantic versioning to its full extent is a good idea for Godot. Also, Godot 3.2.4 isn't going to be versioned as Godot 3.3 because we'd then have 3 different branches the maintain at the same time (`3.2`, `3.3` and `master`).

However, we discussed that after 4.0, we should switch to timed release schedules for patch releases (such as making a new release every 6 weeks, like web browsers or Rust).

**Comment by golddotasksquestions** [comment]
> Godot 3.2.4 isn't going to be versioned as Godot 3.3 because we'd then have 3 different branches the maintain at the same time (3.2, 3.3 and master).

Thanks for the lightning fast response! :) 
I have heard about this argument before, but I don't understand it. 
Why would there have to be 3 different branches maintained? 3.2 could also just receive ["Only critical, security and platform support fixes."](https://docs.godotengine.org/en/stable/about/release_policy.html) just like 3.1, 3.0, and 2.1, while the Godot 3 branch is kept alive receiving LTS with 3.3.X

**Comment by mrjustaguy** [comment]
The issue is that there'd be tons of branches to keep up with, as the minors would be added quite often and currently branches are Major.Minor, not just Major. Might make sense to change with 4 going forward.

**Comment by golddotasksquestions** [comment]
> The issue is that there'd be tons of branches to keep up with

I don't quite understand this argument. When was the latest 3.0 or 3.1 update? I don't remember any updates to be honest. How would it be different going forward if features to the Godot 3 and Godot 4 branch would only added in minor and major versions?

**Comment by Xrayez** [comment]
*Disclaimer: That's my personal opinion and the way I see the situation.*

I sympathize with desire to have more "stable" version of Godot. In my opinion, PRs like godotengine/godot#45618 shouldn't be backported (even if they're donated by IMVU). 😛

I was even told that **future-proofing** is discouraged for Godot development, but godotengine/godot#45618 proves otherwise.

We just end up with broken releases which are unlikely to be fixed. Maybe most apparent issues will be fixed eventually, but what about the other issues which are left unnoticed because of those changes? Experienced developers are more likely than not to switch to Godot 4.0 and dump 3.2 eventually: https://github.com/godotengine/godot-proposals/issues/1533#issuecomment-699557023.

> The 3.2 branch will eventually be phased out like what happened with the previous versions.

So, I totally understand the problem this proposal is trying to pin-point. Unfortunately, I think there's no easy solution to this, because there are some (external) factors that won't allow this:
1. Lack of people maintaining Godot (maybe there are capable people, but I see no delegation regarding project management).
2. Projects which use Godot 3.2 commercially (or as they call "serious projects"). Usually they *have* to use Godot 3.2 in production, but then we see how things are left unimplemented and/or broken in 3.2, but then again, most core developers switched to Godot 4.0 even under those circumstances. Yet developers choose to contribute to Godot 3.2 even if those changes have a high chance to introduce regressions.

What I'm trying to say: I think that Godot 3.x "supposed" to stay as a major release for some time (so we'd definitely see 3.3 and so on), but core devs suddenly decided to work on Vulkan in 4.0 in a separate branch (again, not sure whether the [Epic's Mega Grant](https://godotengine.org/article/godot-engine-was-awarded-epic-megagrant) had any influence on bumping the major version), which made the gap quite big, and rebasing the `vulkan` branch has become difficult with each change to 3.2, so they decided to merge it to `master`. Some users/devs cannot even launch Godot 4.0, but they still want to contribute. That's the source of those breaking changes, regressions etc. in 3.2 (and of course backward-compatible features/enhancements).

**Comment by akien-mga** [comment]
For the reference, what you describe is basically Godot's release policy until 3.2.

For 3.2.x, we initially decided that it would be the last stable branch before 4.0, to avoid having two development branches in parallel and duplicating effort. But user (and contributor) pressure to get more features in 3.2 led us to actually include significant changes in 3.2.x releases, especially 3.2.2 and the upcoming 3.2.4.

We've finally decided to fix that and to rename 3.2.4 to 3.3, and therefore open a `3.x` development branch for future minor releases.

So from now on, Godot's release policy should go back to how it was before 3.2 and that's basically what you describe in the OP. That's not a strict semantic versioning, as if we were strict, we would bump minor version every other commit (each time we add new APIs) and mostly likely bump major version all the time too (a bugfix can make a given feature change behavior in a way that can be considered not backward compatible).

I've described the release policy that I plan to follow for 3.3+, which is basically along the same lines as this proposal (I hadn't actually read this proposal before now, but I'm happy to see it seems to match my own :)):
https://github.com/godotengine/godot-docs/pull/4757
